### DISCUSSION TITLE
Passing data from `script setup` to nitro hooks.

### DISCUSSION THREAD

**Discussion by madc** [nuxt/nuxt#27113] [Questions] [answered]
I'm working on a module to be used with static websites (`generate`).
 
As a basic concept, on `generate` I want to collect data specific to each. The information is already available in the pages `script setup`, so my thought was to collect it there using a composable (i.e. `useCollectData`). This part works as expected.

Where I'm having problems is, how to pass this data to the nitro `prerender:done` hook and write them to a file in `publicDir`. In nuxt 2 we used `ssrContext` to pass on the information .
I'd be thankful for any hint on how to do the save this in nuxt 3.

Thanks.

**Comment by TheAlexLichter** [comment, marked as answer]
if you get the data on the server side, use `useStorage` (unstorage) and save them to the FS 😋

**Reply by madc** [reply]
I was focused on getting this done in memory, but you're right, filesystem could work. But I'd asume `useStorage` is not available in nitro hooks, is there an elegant way to fetch the data?

**Reply by TheAlexLichter** [reply]
Hmm, not sure if it isn't, it might be. Otherwise plain `createStorage` from unstorage :)

**Reply by madc** [reply]
Based on your comments I was able to create a working prove of concept:

**nuxt.config.ts**:
```
export default defineNuxtConfig({
// [...]
  nitro: {
    storage: {
      db: {
        driver: 'fs',
        base: './.data',
      },
    },
  },
// [...]
})
```
In my composable I am now able to do this:
```
export function useMyComposable() {
// [...]
    if (process.server) {
      await useStorage().setItem('db:collection', collection)
    }
// [...]
}
```

And in my nitro hook I am able to fetch the data (I could not figure out, how to import `useStorage` at this point.):
```
export default defineNuxtModule({
// [...]
  setup(options, nuxt) {
    nuxt.hooks.hook('nitro:init', async (nitro) => {
      nitro.hooks.hook('prerender:done', async () => {
        const collection = (await nitro.storage.getItem('db:collection')) || {}
      })
    })
  },
})
```

It is not perfect yet, but it works. Always open for suggestions to improve.

**Comment by madc** [comment]
There is no way to extend the `route` object inside the `prerender:generate` nitro hook from setup, is there? This would be a nice way to pass data without having to temporary persist them.
